### PR TITLE
subcommands

### PR DESCRIPTION
**Pull request by kzidane**
E.g.,

```
ide50 start ...
ide50 stop
ide50 status
...
```

### CODE REVIEW

File: ide50/__main__.py [hunk 1]
```diff
@@ -23,53 +25,82 @@
...
         sys.exit(_("Docker not running"))
     except subprocess.TimeoutExpired:
         sys.exit(_("Docker not responding"))
 
+
+def container_info(container):
+    """Print container port mappings and URL"""
+    # List port mappings
+    print(ports(container))
+    print(_("Running on {}. Run `ide50 -S` to stop.").format(
```
Comment (dmalan) [inline ide50/__main__.py:96]: `ide50 stop`?

File: ide50/__main__.py [hunk 2]
```diff
@@ -78,32 +109,61 @@ def main():
...
+        digest = subprocess.check_output(["docker", "inspect", "--format",
+                                          "{{index .RepoDigests 0}}", image],
+                                         stderr=subprocess.DEVNULL).decode().rstrip()
+
+        # Get digest of latest image
+        # https://stackoverflow.com/a/50945459/5156190
+        repository, tag = image.split(":") if ":" in image else (image, "latest")
+        if "/" not in repository:
+            repository = "library/" + repository
+
```
Comment (dmalan) [inline ide50/__main__.py:135]: unnecessary blank line?

File: ide50/__main__.py [hunk 3]
```diff
@@ -78,32 +109,61 @@ def main():
...
+        try:
+            latest = max(requests.get("https://pypi.org/pypi/ide50/json").json()["releases"],
+                         key=pkg_resources.parse_version)
+
+            assert latest <= __version__
+        except (json.decoder.JSONDecodeError, requests.RequestException):
+            pass
+        except AssertionError:
+            print(_("A newer version is available. Run `pip3 install --upgrade ide50` to upgrade."))
+
```
Comment (dmalan) [inline ide50/__main__.py:160]: comment?

File: ide50/__main__.py [hunk 4]
```diff
@@ -142,50 +204,48 @@ def main():
             ports_ = [f"--publish={port}:{port}" for port in (C9_PORT, 8080, 8081, 8082)]
             container = subprocess.check_output([*cmd, *ports_, args.image],
                                                 stderr=subprocess.STDOUT).decode().rstrip()
-
         except subprocess.CalledProcessError:
 
             # Publish all exposed ports to random ports
             try:
-                container = subprocess.check_output([*cmd, "--publish-all", args.image]).decode().rstrip()
+                container = subprocess.check_output([*cmd, "--publish-all",
```
Comment (dmalan) [inline ide50/__main__.py:211]: clever

File: ide50/__main__.py [hunk 5]
```diff
@@ -142,50 +204,48 @@ def main():
...
+                                                     args.image]).decode().rstrip()
+
             except subprocess.CalledProcessError:
                 sys.exit(_("Failed to start container"))
 
-    # List port mappings
-    print(ports(container))
-    print(_("Running on {}. Run `ide50 -S` to stop.").format(f"http://{ports(container.split()[0], C9_PORT)}/"))
-
+    container_info(container)
```
Comment (dmalan) [inline ide50/__main__.py:217]: comment?

File: ide50/__main__.py [hunk 6]
```diff
@@ -142,50 +204,48 @@ def main():
...
+    container = get_container()
+    if container:
+        container_info(container)
+    else:
+        print(_("No containers are running"))
 
-def pull(image):
-    """Pull image as needed."""
-    try:
+    sys.exit(0)
```
Comment (dmalan) [inline ide50/__main__.py:228]: unnecessary blank line?

**Review by dmalan — CHANGES_REQUESTED** [review]

File: ide50/__main__.py (hunk 2 above)
Comment (kzidane) [reply ide50/__main__.py:135]: I typically use a blank line after `if`s, `for`s, groups of related statements, multi-line statements, etc to visually separate them https://www.python.org/dev/peps/pep-0008/#blank-lines.

File: ide50/__main__.py (hunk 1 above)
Comment (kzidane) [reply ide50/__main__.py:96]: Good catch! Fixed. Thanks!

**Pull request merged**